### DISCUSSION TITLE
Replace magic comments with SET

### DISCUSSION THREAD

**Discussion by kyleconroy** [sqlc-dev/sqlc#871] [Feature Requests & Ideas]
sqlc currently uses a magic comment to configure per-query name and return type:

```sql
-- name: GetAuthor :one
SELECT * FROM authors
WHERE id = $1 LIMIT 1;
```

While this has worked well so far, there are some significant downsides. It's very easy to make [a typo](https://github.com/kyleconroy/sqlc/pull/859). If we want to add additional per-query configuration options ([custom models](https://github.com/kyleconroy/sqlc/issues/755), overrides, etc.), we need to change the syntax.

Instead of continuing down the comment path, I think it's better to use the `SET` command, already part of the SQL standard.

```sql
SET query_name = 'GetAuthor';
SET go_command = 'one';
SELECT * FROM authors
WHERE id = $1 LIMIT 1;
```

Thoughts?

https://www.postgresql.org/docs/current/sql-set.html
https://dev.mysql.com/doc/refman/8.0/en/set-statement.html

**Comment by kevinburke1** [comment]
I like the conciseness of the magic comment. I'd be nervous about making it more difficult to write a query. I hope you'd continue supporting that and maybe also support other methods if you need more complexity.

**Reply by kyleconroy** [reply]
> I hope you'd continue supporting that 

Yes! The magic comment support will not be going away, as that would break a bunch of existing code.

**Reply by kevinburke1** [reply]
Ok, that's good. Also I guess the idea of assigning a variable and then re-assigning it to different values 20 times in the same file just seems icky to me.

**Comment by kyleconroy** [comment]
`PREPARE` might actually be a better fit.

```sql
CREATE TABLE authors (
  id   BIGSERIAL PRIMARY KEY,
  name text      NOT NULL,
  bio  text
);

PREPARE get_author AS
  SELECT * FROM authors
  WHERE id = $1 LIMIT 1;

PREPARE list_authors AS
  SELECT * FROM authors
  ORDER BY name;

PREPARE create_author AS
  INSERT INTO authors (
    name, bio
  ) VALUES (
    $1, $2
  )
  RETURNING *;

PREPARE delete_authors AS
  DELETE FROM authors
  WHERE id = $1;
```
